### PR TITLE
Add signup and recover possibility

### PR DESCRIPTION
**Pull request by ineiti**
Use the new email part of the personhood service that allows to:
- sign up a new user and send an email with the sign-up link
- recover an existing user with an existing email

### CODE REVIEW

File: webapp/src/app/newuser/newuser.component.ts [hunk 1]
```diff
@@ -1,7 +1,102 @@
...
-  templateUrl: "./newuser.component.html",
+    selector: "app-newuser",
+    templateUrl: "./newuser.component.html",
 })
-export class NewuserComponent {}
+export class NewuserComponent implements OnInit {
+    signupForm: FormGroup;
+    recoveryForm: FormGroup;
+
+    constructor(private snack: MatSnackBar,
```
Comment (tharvik) [inline webapp/src/app/newuser/newuser.component.ts:19]: `snack` is unused

File: webapp/src/app/newuser/newuser.component.ts [hunk 2]
```diff
@@ -1,7 +1,102 @@
...
+                private dialog: MatDialog,
+                public bcs: ByzCoinService) {
+    }
+
+    async ngOnInit() {
+        for (const node of this.bcs.config.roster.list) {
+            Log.info("node", node.description, "id", node.id);
+        }
```
Comment (tharvik) [inline webapp/src/app/newuser/newuser.component.ts:27]: debug?

File: webapp/src/app/newuser/newuser.component.ts [hunk 3]
```diff
@@ -1,7 +1,102 @@
...
+        this.signupForm = new FormGroup({
+            alias: new FormControl(),
+            email: new FormControl("", Validators.email),
+        });
+        this.recoveryForm = new FormGroup({
+            email: new FormControl("", Validators.email),
+        });
```
Comment (tharvik) [inline webapp/src/app/newuser/newuser.component.ts:34]: you can move theses to the constructor and have the fields readonly

File: webapp/src/assets/config.dedis.toml [hunk 4]
```diff
@@ -1,4 +1,5 @@
 ByzCoinID = "9cc36071ccb902a1de7e0d21a2c176d73894b1cf88ae4cc2ba4c95cd76f474f3"
+SignupNode = "https://byzcoin-signup.c4dt.org"
```
Comment (tharvik) [inline webapp/src/assets/config.dedis.toml:2]: not really related to the PR; `curl http://byzcoin-signup.c4dt.org` (via http not https) gives me a 404, shouldn't it redirect to the https version instead?

File: webapp/src/lib/config.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import { Roster } from "@dedis/cothority/network/proto";
 
 import toml from "toml";
+import Log from "@dedis/cothority/log";
```
Comment (tharvik) [inline webapp/src/lib/config.ts:4]: unused

File: webapp/src/lib/config.ts [hunk 6]
```diff
@@ -26,7 +27,17 @@ export class Config {
...
                 throw Error("is not a string");
             }
-            if (!(/[a-f0-9]{64}/).test(field)) {
+            if (!(/^[a-f0-9]{64}$/).test(field)) {
+                throw Error("is not of correct format");
+            }
+            return Buffer.from(field, "hex");
+        };
+
+        const asUUID = (field: any): ID => {
```
Comment (tharvik) [inline webapp/src/lib/config.ts:36]: unused

File: webapp/src/app/newuser/newuser.component.ts [hunk 7]
```diff
@@ -1,7 +1,102 @@
...
+    }
+
+    getPHRPC(): PersonhoodRPC {
+        if (this.bcs.config.signupNode === "") {
```
Comment (tharvik) [inline webapp/src/app/newuser/newuser.component.ts:38]: the field is `undefined` if it is not existing in the config, not the empty string

```suggestion
        if (this.bcs.config.signupNode === undefined) {
```

File: webapp/src/lib/config.ts [hunk 8]
```diff
@@ -43,6 +54,7 @@ export class Config {
         return new Config(
             getField("ByzCoinID", asID),
             Roster.fromTOML(raw),
+            tryToGetField("SignupNode", asString),
```
Comment (tharvik) [inline webapp/src/lib/config.ts:57]: consider parsing it as URL to ensure that it is in the correct format and fails at load time, not latter

**Review by tharvik — CHANGES_REQUESTED** [review]
some cosmetics (I looove `private readonly`) and a bug when missing the `signupNode` from the config

File: webapp/src/lib/config.ts (hunk 6 above)
Comment (ineiti) [reply webapp/src/lib/config.ts:36]: removed

File: webapp/src/lib/config.ts (hunk 8 above)
Comment (ineiti) [reply webapp/src/lib/config.ts:57]: Added `asURL`, even though for the `ServerIdentity` I need to convert it back to a string...

Comment (tharvik) [reply webapp/src/lib/config.ts]: yep, ideally, `ServerIdentity` should take a URL, but for now, at least the configuration will be correct.

I remember that some times ago, I already wanted to change cothority to use the correct URL type, to catch misconfiguration early, but it was cumbersome as it wasn't working well with protobuf.

**Review by tharvik — APPROVED** [review]

**Pull request merged**